wdl-lint/src/rules/shellcheck.rs

```diff
@@ -375,13 +378,98 @@ impl<'a> CommandContext<'a> {
     }
 }
 
+/// Detect embedded quotes surrounding an expression in a string.
+///
+/// This is a utility function called by `evaluates_to_literal`. Only `expr`
+/// that are addition or strings with potentially embedded placeholders are
+/// valid input. For a given expression, it checks through all descendants to
+/// see if there are any name references (variables) that are surrounded by
+/// escaped quotes. In WDL, the parent expression is either an addition
+/// (concatenation, e.g. `~{"foo " + bar + " baz"}`) operation or a string with
+/// an embedded placeholder (e.g. `~{"foo ~{bar} baz"`). So the escaped quotes
+/// are not in a single string literal. The descendant expressions must be
+/// traversed to check for quoting.
+fn is_quoted(expr: &Expr) -> bool {
+    let mut opened = false;
+    let mut name = false;
+
+    for c in expr.descendants::<Expr>() {
+        match c {
+            Expr::Literal(LiteralExpr::String(ref s)) => {
+                for p in s.parts() {
+                    match p {
+                        StringPart::Text(t) => {
+                            let mut buffer = String::new();
+                            t.unescape_to(&mut buffer);
+                            buffer.match_indices(&['\'', '"']).for_each(|(..)| {
+                                if opened && name {
+                                    name = false;
+                                }
+                                opened = !opened;
+                            });
+                        }
+                        StringPart::Placeholder(_placeholder) => {
+                            if !opened {
+                                return false;
+                            }
+                            name = true;
+                        }
+                    }
+                }
+            }
+            Expr::NameRef(_) => {
+                if !opened {
+                    return false;
+                }
+                name = true;
+            }
+            _ => {}
+        }
+    }
+    !name
+}
+
+/// Evaluate an expression to determine if it can be simplified to a literal.
+fn evaluates_to_literal(expr: &Expr) -> bool {
```

Comment:
```suggestion
/// Evaluate a WDL expression to determine if it can be simplified to a bash literal.
fn evaluates_to_bash_literal(expr: &Expr) -> bool {
```
This is probably obvious to someone more familiar with the codebase, but I spent quite a while scratching my head at this function before realizing we were talking about two different kinds of literals.